keras_cv/bounding_box/iou.py

```diff
@@ -50,16 +95,19 @@ def compute_iou(
     boxes1_rank = len(boxes1.shape)
     boxes2_rank = len(boxes2.shape)
 
-    if (
-        boxes1_rank != boxes2_rank
-        or boxes1_rank not in [2, 3]
-        or boxes2_rank not in [2, 3]
-    ):
+    if boxes1_rank not in [2, 3]:
```

Comment:
perhaps instead of supporting differing ranks we should follow the TensorFlow convention and broadcast only when the first axis is of dim [1].

So instead of:

[boxes, 4], [batch, boxes, 4]

you would pass:

[1, boxes, 4], [batch, boxes, 4]

This is more consistent with the TensorFlow API than automatically broadcasting for users

Reply:
Here the problem is not on [1, boxes, 4] vs [batch, boxes, 4], it's actually [boxes_1, 4] & [batch, boxes_2, 4] should give user [batch, boxes_1, boxes_2, 4], if users pass in [1, boxes_1, 4] and [batch, boxes_2, 4], this will actually error out telling user that the shape doesn't match, which is expected.
If we do ask users to convert to [batch, boxes_1, 4] before calling compute_iou, then users would have to do tf.tile(tf.expand_dims(boxes1, 0), tf.stack([tf.shape(boxes_2)[0], tf.shape(boxes_1)[1:])) in order to call it, which to me is too bloated.

Reply:
I'm no expert in the bbox conventions, but personally I think this API looks good.

Reply:
Sounds good, that’s fine with me.